File: src/string-and-array/course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> ans;
        vector<vector<int>> graph(numCourses);
        vector<int> indegress(numCourses, 0);
        queue<int> q;

        for(auto &prerequisite : prerequisites) {
            graph[prerequisite[1]].push_back(prerequisite[0]);
            ++indegress[prerequisite[0]];
        }

        for(int i = 0; i < numCourses; ++i) {
            if(indegress[i] == 0) {
                q.push(i);
            }
        }

        while(q.size()) {
            int cur = q.front();
            q.pop();
            ans.push_back(cur);

            for(int neighbor : graph[cur]) {
                if(--indegress[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }

        return ans.size() == numCourses ? ans : vector<int>();
    }
};

File: src/graph-and-tree/course-schedule-ii.ts
export function findOrder(
  numCourses: number,
  prerequisites: number[][]
): number[] {
  let ans: number[] = []

  const graph: number[][] = []
  const indegree: number[] = []
  for (let i = 0; i < numCourses; i++) {
    graph[i] = []
    indegree[i] = 0
  }

  const len = prerequisites.length

  for (let i = 0; i < len; i++) {
    const [to, from] = prerequisites[i]

    graph[from].push(to)
    indegree[to]++
  }

  const queue: number[] = []

  for (let i = 0; i < numCourses; i++) {
    if (indegree[i] === 0) {
      queue.push(i)
    }
  }

  while (queue.length) {
    const vertex = queue.pop()!
    ans.push(vertex)
    for (const neighbor of graph[vertex]) {
      if (--indegree[neighbor] === 0) {
        queue.push(neighbor)
      }
    }
  }

  return ans.length === numCourses ? ans : []
}
